swarm_dial.go

```diff
@@ -349,65 +352,44 @@ func (s *Swarm) dialAddrs(ctx context.Context, p peer.ID, remoteAddrs <-chan ma.
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel() // cancel work when we exit func
 
-	// use a single response type instead of errs and conns, reduces complexity *a ton*
-	respch := make(chan dialResult)
+	dialResults := make(chan dialResult)
 
-	defaultDialFail := fmt.Errorf("failed to dial %s (default failure)", p)
-	exitErr := defaultDialFail
+	var lastDialErr error
 
 	defer s.limiter.clearAllPeerDials(p)
 
 	var active int
 	for remoteAddrs != nil || active > 0 {
-		// Check for context cancellations and/or responses first.
-		select {
-		case <-ctx.Done():
-			if exitErr == defaultDialFail {
-				exitErr = ctx.Err()
-			}
-			return nil, exitErr
-		case resp := <-respch:
-			active--
-			if resp.Err != nil {
-				log.Infof("got error on dial to %s: %s", resp.Addr, resp.Err)
-				// Errors are normal, lots of dials will fail
-				exitErr = resp.Err
-			} else if resp.Conn != nil {
-				return resp.Conn, nil
-			}
-
-			// We got a result, try again from the top.
-			continue
-		default:
-		}
-
-		// Now, attempt to dial.
 		select {
 		case addr, ok := <-remoteAddrs:
 			if !ok {
 				remoteAddrs = nil
 				continue
 			}
-
-			s.limitedDial(ctx, p, addr, respch)
+			s.limitedDial(ctx, p, addr, dialResults)
 			active++
 		case <-ctx.Done():
-			if exitErr == defaultDialFail {
-				exitErr = ctx.Err()
+			if lastDialErr != nil {
+				return nil, lastDialErr
+			} else {
+				return nil, ctx.Err()
 			}
-			return nil, exitErr
-		case resp := <-respch:
+		case resp := <-dialResults:
 			active--
 			if resp.Err != nil {
 				log.Infof("got error on dial to %s: %s", resp.Addr, resp.Err)
 				// Errors are normal, lots of dials will fail
-				exitErr = resp.Err
-			} else if resp.Conn != nil {
+				lastDialErr = resp.Err
+			} else {
 				return resp.Conn, nil
 			}
 		}
 	}
-	return nil, exitErr
+	if lastDialErr != nil {
+		return nil, lastDialErr
+	} else {
+		return nil, errors.New("no remote addresses")
```

Comment:
This is a validation error and it should be caught at the beginning for clarity. Also, let's make it an exported error.

Reply:
Actually I think it's better to put it at the end. In this instance it's flexing a constraint of this method: If there are any remote addresses and none of them connect, then by we should have a dial error. Putting it at the end here reduces the special conditions on this method.

Reply:
Superfluous `else`.

Reply:
Sure, this one is a guard.